src/test/java/ca/team2706/frc/robot/commands/drivebase/DriverAssistVisionTest.java

```diff
@@ -0,0 +1,156 @@
+package ca.team2706.frc.robot.commands.drivebase;
+
+import ca.team2706.frc.robot.SendablesTest;
+import ca.team2706.frc.robot.config.Config;
+import ca.team2706.frc.robot.subsystems.DriveBase;
+import com.ctre.phoenix.CTREJNIWrapper;
+import com.ctre.phoenix.motorcontrol.can.MotControllerJNI;
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+import com.ctre.phoenix.sensors.PigeonIMU;
+import edu.wpi.first.wpilibj.AnalogInput;
+import edu.wpi.first.wpilibj.Notifier;
+import edu.wpi.first.wpilibj.PWM;
+import edu.wpi.first.wpilibj.drive.DifferentialDrive;
+import jaci.pathfinder.Pathfinder;
+import jaci.pathfinder.Trajectory;
+import mockit.Expectations;
+import mockit.Mocked;
+import mockit.Tested;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+;
+
+public class DriverAssistVisionTest {
+    @Tested
+    private DriverAssistVision driverAssistVision;
+
+    DriveBase driveBase = DriveBase.getInstance();
+
+    @Mocked
+    private WPI_TalonSRX talon;
+
+    @Mocked
+    private PWM pwm;
+
+    @Mocked
+    private AnalogInput analogInput;
+
+    @Mocked(stubOutClassInitialization = true)
+    private PigeonIMU pigeon;
+
+    private DifferentialDrive differentialDrive;
+
+    @Mocked(stubOutClassInitialization = true)
+    private CTREJNIWrapper jni;
+
+    @Mocked(stubOutClassInitialization = true)
+    private MotControllerJNI motControllerJNI;
+
+    @Mocked
+    private Notifier notifier;
+
+    /**
+     * Runs various test for generating a trajectory to targets on the cargo ship or
+     * the loading bay. For each test, the same vision input is used but the heading
+     * of the robot in the field frame (referred to as "absolute heading" in the code)
+     * is changed.
+     */
+    @Test
+    public void testGenerateTrajectoryCargoShipAndLoading() {
+        new Expectations(Config.class) {{
```

Comment:
Since you're not actually mocking `Config`, these expectations won't actually work. Maybe you could factor those constants into your expected values somehow?

Reply:
Actually @eandr127 this is partial mocking. Basically when you only want to mock the methods that you mention in the expectations, you can do 
```java
new Expectations(class) {{
    class.method();
    result  = something;
}};
```
which **only** mocks `class.method()`, leaving the rest of the methods untouched. 

Since this is still mocking config I still don't think we should have this at all.

Reply:
They should work and I've just confirmed this now by running the test after changing one of the Expectations which caused the test to fail.

Config is now being partially mocked as described in the comment [https://github.com/FRC2706/2019-2706-Robot-Code/pull/99#discussion_r261840421](url)

Reply:
Also, factoring them into the expected values would be quite complicated. The Expectations were chosen so that the final values produced values that could be deduced by relatively simple geometry.

Reply:
I changed the test to not use mocking at all. (This also means that the `get` methods in Config that had to be added for the mocking were removed.) The changes have been pushed to Github.

I nevertheless think that, going forward, there should be some mechanism to be able to write unit tests that assume certain specific parameters in the program. (They need not be in Config but can be in some other class considered mockable.) There are several arguments for this:

1.) It can be sometimes very difficult to write tests that accommodate changes to the parameters. This will be especially true if code is one day written for advanced motion planning or the dynamics of the robot.  

2.) It allows you to write specific tests that allows you check if the code handles certain values of the parameters.

3.) It allows you to write tests that will ensure that the code will get fully exercised in the test. It is possible that somebody might set of certain parameter to zero in the field version of the code that effectively causes a certain part of the code to not be run. Bugs can be inadvertently inserted in that portion of the code and it will not be tested. Then one day the parameter will get switched back on and the problem will show up. It is better to catch the bug early.

Reply:
I agree with this, but I don't think that the answer is mocking. Mocking is meant to replace libraries that are out of your control, in our case sensors. 

What I would instead suggest we try to do is make the code use static methods for logic. For example, instead of 
```java
public int doSomething() {
    return Config.SOME_VALUE * 2;
}
```

we could have the following arrangment:
```java
public void doSomething() {
    return doSomething(Config.SOME_VALUE);
}

private static int doSomething(final int value) {
    return value  * 2;
}
```

The tests for this class could therefore use the static method to test logic with a defined parameter.